src/core/raster/qgsrastercontourrenderer.cpp

```diff
@@ -0,0 +1,226 @@
+{
+  if ( parentElem.isNull() )
+  {
+    return;
+  }
+
+  QDomElement rasterRendererElem = doc.createElement( QStringLiteral( "rasterrenderer" ) );
+  _writeXml( doc, rasterRendererElem );
+
+  rasterRendererElem.setAttribute( QStringLiteral( "band" ), mInputBand );
+  rasterRendererElem.setAttribute( QStringLiteral( "contour-interval" ), mContourInterval );
+  rasterRendererElem.setAttribute( QStringLiteral( "contour-index-interval" ), mContourIndexInterval );
+  rasterRendererElem.setAttribute( QStringLiteral( "downscale" ), mDownscale );
+
+  QgsSymbolMap symbols;
+  symbols[QStringLiteral( "contour" )] = mContourSymbol.get();
+  if ( mContourIndexSymbol )
+    symbols[QStringLiteral( "index-contour" )] = mContourIndexSymbol.get();
+  QDomElement symbolsElem = QgsSymbolLayerUtils::saveSymbols( symbols, QStringLiteral( "symbols" ), doc, QgsReadWriteContext() );
+  rasterRendererElem.appendChild( symbolsElem );
+
+  parentElem.appendChild( rasterRendererElem );
+}
+
+struct ContourWriterData
+{
+  QPainter *painter;
+  double scaleX, scaleY;
+  QgsLineSymbol *symbol;
+  QgsLineSymbol *indexSymbol;
+  double indexInterval;
+  QgsRenderContext *context;
+};
+
+CPLErr _rasterContourWriter( double dfLevel, int nPoints, double *padfX, double *padfY, void *ptr )
+{
+  Q_UNUSED( dfLevel )
+  ContourWriterData *crData = static_cast<ContourWriterData *>( ptr );
+  QPolygonF polygon( nPoints );
+  QPointF *d = polygon.data();
+  for ( int i = 0; i < nPoints; ++i )
+  {
+    d[i] = QPointF( padfX[i] * crData->scaleX, padfY[i] * crData->scaleY );
+  }
+
+  if ( crData->indexSymbol && !qgsDoubleNear( crData->indexInterval, 0 ) && qgsDoubleNear( fmod( dfLevel, crData->indexInterval ), 0 ) )
+    crData->indexSymbol->renderPolyline( polygon, nullptr, *crData->context );
+  else
+    crData->symbol->renderPolyline( polygon, nullptr, *crData->context );
+  return CE_None;
+}
+
+QgsRasterBlock *QgsRasterContourRenderer::block( int bandNo, const QgsRectangle &extent, int width, int height, QgsRasterBlockFeedback *feedback )
+{
+  Q_UNUSED( bandNo )
+
+  std::unique_ptr< QgsRasterBlock > outputBlock( new QgsRasterBlock() );
+  if ( !mInput || !mContourSymbol )
+  {
+    return outputBlock.release();
+  }
+
+  int inputWidth = static_cast<int>( round( width / mDownscale ) );
+  int inputHeight = static_cast<int>( round( height / mDownscale ) );
+
+  std::shared_ptr< QgsRasterBlock > inputBlock( mInput->block( mInputBand, extent, inputWidth, inputHeight, feedback ) );
+  if ( !inputBlock || inputBlock->isEmpty() )
+  {
+    QgsDebugMsg( QStringLiteral( "No raster data!" ) );
+    return outputBlock.release();
+  }
+
+  if ( !inputBlock->convert( Qgis::Float64 ) ) // contouring algorithm requires double
+    return outputBlock.release();
+  double *scanline = reinterpret_cast<double *>( inputBlock->bits() );
+
+  QImage img( width, height, QImage::Format_ARGB32_Premultiplied );
```

Comment:
Hmm this is kinda ugly -- it's revealing a shortcoming in the raster renderer. Does it work ok with print layouts and hi-dpi outputs? 

(A pretty-gross-but-api-conserving-fix might be to add the original render context to QgsRasterBlockFeedback)

Reply:
Are you referring to the fact that all the rendered lines get rasterized? That is sub-optimal for print output unfortunately, but I can't see a way around it without rewriting a chunk of the raster pipeline :-) It would be a good exercise to make a list of things that are not ideal in the current raster pipeline implementation and see how we could change the design to fix them...

I have tried with a print layout and things look okay. I have seen some small discontinuities in contours at the edges of blocks created by QgsRasterDrawer, but that can be fixed at some point later I guess.

Reply:
Are you against my temperorary hack of adding a render context to the feedback?

Reply:
Do you mean to add render context to QgsRasterBlockFeedback? I guess I won't mind...

Reply:
Yeah. It's an ugly hack but avoids API breakage. For 4.0 we can make a QgsRasterBlockContext or similar to clean this up...

Reply:
Sounds good.

Raster block context would make a lot of sense, especially for raster renderers which needlessly create new temporary QImages just to be copied to the final render target. The world would become a better place if we won't do that :grin:  (easier to show intermediate results of raster rendering etc.)